src/application.ts

```diff
@@ -0,0 +1,148 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {logger as defaultLogger, Logger} from './logger';
+import {Octokit} from '@octokit/rest';
+import {GitHubContext, Credentials, Parameters} from './types';
+import {runGitHub} from './handle-github';
+import {makeCLIChanges} from './get-changes';
+import * as yargs from 'yargs';
+import {execSync} from 'child_process';
+
+interface GitHubSuggester {
+  accessToken: string;
+  writePermissions: boolean;
+  octokit: Octokit;
+  log: Logger;
+  gitHubContext: GitHubContext;
+  run(): void;
+}
+
+class CLISuggester implements GitHubSuggester {
+  static readonly outputDir = 'out';
+  accessToken!: string;
+  writePermissions: boolean;
+  octokit!: Octokit;
+  log!: Logger;
+  gitHubContext!: GitHubContext;
+
+  constructor(params: Parameters) {
+    this.writePermissions = params.writePermissions || false;
+    this.setupLogger(params.logger);
+    this.setupAuthentication();
+    this.setupGitHubContext();
```

Comment:
I think arg parsing logic actually belongs in the CLI script itself. It should grab all the inputs and provide them to the GitHubSuggester as configuration parameters. The core GitHubSuggester should essentially only handle what's in the `runGitHub()` call with all the configuration passed in the constructor.

For example, this constructor should take in the GitHubContext as a parameter, and the src/bin/code-suggester should parse the CLI inputs to populate those fields.